zcash_client_backend/src/data_api.rs

```diff
@@ -0,0 +1,583 @@
+//! Interfaces for wallet data persistence & low-level wallet utilities.
+
+use std::cmp;
+use std::collections::HashMap;
+use std::fmt::Debug;
+
+use zcash_primitives::{
+    block::BlockHash,
+    consensus::BlockHeight,
+    merkle_tree::{CommitmentTree, IncrementalWitness},
+    note_encryption::Memo,
+    primitives::{Note, Nullifier, PaymentAddress},
+    sapling::Node,
+    transaction::{components::Amount, Transaction, TxId},
+    zip32::ExtendedFullViewingKey,
+};
+
+use crate::{
+    address::RecipientAddress,
+    data_api::wallet::ANCHOR_OFFSET,
+    decrypt::DecryptedOutput,
+    proto::compact_formats::CompactBlock,
+    wallet::{AccountId, SpendableNote, WalletShieldedOutput, WalletTx},
+};
+
+pub mod chain;
+pub mod error;
+pub mod wallet;
+
+/// Read-only operations require for light wallet functions.
+///
+/// This trait defines the read-only portion of the storage
+/// interface atop which higher-level wallet operations are
+/// implemented. It serves to allow wallet functions to be
+/// abstracted away from any particular data storage substrate.
+pub trait WalletRead {
+    /// The type of errors produced by a wallet backend.
+    type Error;
+
+    /// Backend-specific note identifier.
+    ///
+    /// For example, this might be a database identifier type
+    /// or a UUID.
+    type NoteRef: Copy + Debug;
+
+    /// Backend-specific transaction identifier.
+    ///
+    /// For example, this might be a database identifier type
+    /// or a TxId if the backend is able to support that type
+    /// directly.
+    type TxRef: Copy + Debug;
+
+    /// Returns the minimum and maximum block heights for stored blocks.
+    fn block_height_extrema(&self) -> Result<Option<(BlockHeight, BlockHeight)>, Self::Error>;
+
+    /// Returns the default target height and anchor height, given the
+    /// range of block heights that the backend knows about.
+    fn get_target_and_anchor_heights(
+        &self,
+    ) -> Result<Option<(BlockHeight, BlockHeight)>, Self::Error> {
+        self.block_height_extrema().map(|heights| {
+            heights.map(|(min_height, max_height)| {
+                let target_height = max_height + 1;
+
+                // Select an anchor ANCHOR_OFFSET back from the target block,
+                // unless that would be before the earliest block we have.
```

Comment:
Should we instead not allow transfers until it is possible to choose an anchor `ANCHOR_OFFSET` blocks back? That condition shouldn't happen very often, but it seems safer to always enforce the same anchor policy (and the caller must already handle errors).

Reply:
I think you're right. In practice though, I think this wouldn't occur because the earliest block in the database wouldn't be spendable until at least ANCHOR_OFFSET blocks have elapsed?

Reply:
The anchor offset isn't a consensus rule, so why wouldn't it be?

Reply:
My reasoning here was that when a wallet is initialized with a birthday height that it won't find any inbound notes below that height, and that the wallet's internal number of confirmation checks will have to have elapsed before it allows those notes to be spent. But I guess that it's possible that the wallet permits fewer than ANCHOR_OFFSET confirmations? I'm not entirely confident about how choices that a wallet is able to make interacts with having this constant set.